src/script/interpreter.cpp

```diff
@@ -1110,35 +1110,46 @@ class CTransactionSignatureSerializer {
 
 } // anon namespace
 
-uint256 SignatureHash(const CScript& scriptCode, const CTransaction& txTo, unsigned int nIn, int nHashType, const CAmount& amount, int sigversion)
+uint256 SignatureHash(const CScript& scriptCode, const CTransaction& txTo, unsigned int nIn, int nHashType, const CAmount& amount, int sigversion, CachedHashes* cache)
 {
+    CachedHashes empty;
+    if(cache == NULL)
+    {
+        cache = &empty;
+    }
+    
     if (sigversion == 1) {
+        if(cache->IsEmpty())
```

Comment:
Other than simplicity why is the computation moved here? If all inputs are ANYONECANPAY then you don't need hashPrevouts, for example. Computing only when necessary doesn't seem overly complex.